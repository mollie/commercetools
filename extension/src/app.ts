import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
