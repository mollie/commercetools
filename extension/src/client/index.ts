import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
