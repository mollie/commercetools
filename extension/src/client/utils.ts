import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
