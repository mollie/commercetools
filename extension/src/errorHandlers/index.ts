import { formatMollieErrorResponse } from './formatMollieErrorResponse';
import { formatExtensionErrorResponse } from './formatExtensionErrorResponse';

const errorHandler = {
  mollie: formatMollieErrorResponse,
  extension: formatExtensionErrorResponse,
};

export default errorHandler;
