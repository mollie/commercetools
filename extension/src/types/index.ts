// This is not exhaustive
// If you use another commercetools error response code, add it to this enum
export enum CTEnumErrors {
  General = 'General',
  InvalidInput = 'InvalidInput',
  InvalidOperation = 'InvalidOperation',
  Unauthorized = 'Unauthorized',
  SyntaxError = 'SyntaxError',
  SemanticError = 'SemanticError',
  ObjectNotFound = 'ObjectNotFound',
}

export type CTUpdatesRequestedResponse = {
  status: number;
  actions?: Action[];
  errors?: CTError[];
};

export type Action = {
  action: string;
  type?: {
    key: string;
  };
  fields?: {
    actionType: string;
    request?: string;
    response?: string;
    createdAt?: string;
  };
  name?: string;
  value?: string;
  key?: string;
  transaction?: CTTransaction;
  transactionId?: string;
  interactionId?: string;
  state?: CTTransactionState;
};

export type CTError = {
  code: CTEnumErrors;
  message: string;
  extensionExtraInfo?: Object;
};

export type CTMoney = {
  type?: 'centPrecision';
  currencyCode: string;
  centAmount: number;
  fractionDigits?: number;
};

export type CTPayment = {
  id: string;
  amountPlanned: CTMoney;
  transactions?: CTTransaction[];
  key?: string;
  custom?: {
    fields?: {
      paymentMethodsRequest?: string;
      paymentMethodsResponse?: string;
      createPayment?: string;
      createCapture?: string;
      cancelPayment?: string;
      createRefund?: string;
    };
  };
  interfaceInteractions?: CTInterfaceInteraction[];
};

export type CTInterfaceInteraction = {
  id: string;
  actionType: string;
  createdAt?: Date;
  request?: string;
  response?: string;
};

export type CTTransaction = {
  id?: string;
  timestamp?: string;
  type: CTTransactionType;
  amount: CTMoney;
  interactionId?: string;
  state?: 'Initial' | 'Pending' | 'Success' | 'Failure';
};

export enum ControllerAction {
  GetPaymentMethods = 'getPaymentMethods',
  CreateOrder = 'createOrder',
  CreateOrderPayment = 'createOrderPayment',
  CreateOrderRefund = 'createOrderRefund',
  CreateCustomRefund = 'createCustomRefund',
  CreateShipment = 'createShipment',
  UpdateShipment = 'updateShipment',
  CancelOrder = 'cancelOrder',
  NoAction = 'noAction',
  Error = 'error',
}

export enum CTTransactionType {
  Authorization = 'Authorization',
  CancelAuthorization = 'CancelAuthorization',
  Charge = 'Charge',
  Refund = 'Refund',
  Chargeback = 'Chargeback',
}

export enum CTTransactionState {
  Initial = 'Initial',
  Pending = 'Pending',
  Success = 'Success',
  Failure = 'Failure',
}
