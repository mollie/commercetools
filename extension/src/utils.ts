import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
