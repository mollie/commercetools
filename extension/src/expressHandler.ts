import {Request, Response} from 'express';
import handleRequest from './requestHandlers/handleRequest';
import {HandleRequestInput, HandleRequestSuccess} from './types';
import actions from "./requestHandlers/actions";


export default async function handler(req: Request, res: Response) {
    const requestInput = new HandleRequestInput(req.path, req.method, req.body);
    const result = await handleRequest(requestInput);
    if (result instanceof HandleRequestSuccess) {
        return res.status(200).send({actions: actions});

    } else {
        return res.status(result.status).send({errors: result.errors});
    }
};
