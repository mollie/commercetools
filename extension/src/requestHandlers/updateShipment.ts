import { MollieClient, ShipmentUpdateParams, Shipment } from '@mollie/api-client';
import { formatMollieErrorResponse } from '../errorHandlers/formatMollieErrorResponse';
import { Action, ControllerAction, CTUpdatesRequestedResponse } from '../types';
import { createDateNowString } from '../utils';
import Logger from '../../src/logger/logger';

export function getShipmentParams(ctObj: any): Promise<{ shipmentId: string; updateParams: ShipmentUpdateParams }> {
  try {
    const parsedShipmentRequest = JSON.parse(ctObj?.custom?.fields?.updateShipmentRequest);
    const shipmentId = parsedShipmentRequest.shipmentId;
    const updateParams: ShipmentUpdateParams = {
      orderId: ctObj.key,
      tracking: parsedShipmentRequest.tracking,
    };

    Logger.debug({ shipmentId: shipmentId });
    Logger.debug({ updateShipmentParams: updateParams });
    return Promise.resolve({ shipmentId, updateParams });
  } catch (error) {
    Logger.error({ error });
    return Promise.reject({ status: 400, title: 'Could not make parameters needed to update Mollie shipment.', field: 'updateShipmentRequest' });
  }
}

export function createCtActions(mollieShipmentRes: Shipment, ctObj: any): Action[] {
  const stringifiedShipmentResponse = JSON.stringify(mollieShipmentRes);
  const result: Action[] = [
    {
      action: 'addInterfaceInteraction',
      type: {
        key: 'ct-mollie-integration-interface-interaction-type',
      },
      fields: {
        actionType: ControllerAction.UpdateShipment,
        createdAt: createDateNowString(),
        request: ctObj?.custom?.fields?.updateShipmentRequest,
        response: stringifiedShipmentResponse,
      },
    },
    {
      action: 'setCustomField',
      name: 'updateShipmentResponse',
      value: stringifiedShipmentResponse,
    },
  ];
  return result;
}

export default async function updateShipment(ctObj: any, mollieClient: MollieClient, getShipmentParams: Function, createCtActions: Function): Promise<CTUpdatesRequestedResponse> {
  try {
    const shipmentParams = await getShipmentParams(ctObj);
    const mollieShipmentRes = await mollieClient.orders_shipments.update(shipmentParams.shipmentId, shipmentParams.updateParams);
    Logger.debug({ mollieShipmentRes: mollieShipmentRes });
    const ctActions = createCtActions(mollieShipmentRes, ctObj);
    return {
      actions: ctActions,
      status: 200,
    };
  } catch (error: any) {
    const errorResponse = formatMollieErrorResponse(error);
    return errorResponse;
  }
}
