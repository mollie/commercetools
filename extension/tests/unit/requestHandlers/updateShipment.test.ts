import { mocked } from 'ts-jest/utils';
import { Action } from '../../../src/types';
import updateShipment, { getShipmentParams, createCtActions } from '../../../src/requestHandlers/updateShipment';
import { createDateNowString } from '../../../src/utils';
import Logger from '../../../src/logger/logger';

jest.mock('../../../src/utils');

describe('getShipmentParams', () => {
  const mockLogError = jest.fn();
  beforeEach(() => {
    Logger.error = mockLogError;
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should create all params for mollie updateShipment call', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          updateShipmentRequest: '{"shipmentId":"shp_dhq5l7","tracking":{"carrier":"PostNL","code":"3SKABA000000000","url":"http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C"}}',
        },
      },
    };
    const expectedUpdateShipmentParams = {
      shipmentId: 'shp_dhq5l7',
      updateParams: {
        orderId: 'ord_3uwvfd',
        tracking: {
          carrier: 'PostNL',
          code: '3SKABA000000000',
          url: 'http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C',
        },
      },
    };
    await expect(getShipmentParams(mockedCtObj)).resolves.toEqual(expectedUpdateShipmentParams);
  });
  it('Should return 400 and error message if creating the parameters throws an error', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          updateShipmentRequest: '',
        },
      },
    };
    const expectedRejectedValue = {
      status: 400,
      title: 'Could not make parameters needed to update Mollie shipment.',
      field: 'updateShipmentRequest',
    };
    await expect(getShipmentParams(mockedCtObj)).rejects.toEqual(expectedRejectedValue);
    expect(mockLogError).toHaveBeenCalledTimes(1);
  });
});

describe('createCtActions', () => {
  beforeEach(() => {
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should create correct ct actions from request and mollies payment', () => {
    const mockedCtObject = {
      custom: {
        fields: {
          updateShipmentRequest: '{"shipmentId":"shp_dhq5l7","tracking":{"carrier":"PostNL","code":"3SKABA000000000","url":"http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C"}}',
        },
      },
    };
    const mockedShipmentResponse: any = {
      resource: 'shipment',
      id: 'shp_dhq5l7',
      orderId: 'ord_3uwvfd',
      createdAt: '2021-10-28T14:30:25+00:00',
      tracking: {
        carrier: 'PostNL',
        code: '3SKABA000000000',
        url: 'http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C',
      },
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.juguf1',
          orderId: 'ord_3uwvfd',
          name: 'orange',
          sku: null,
          type: 'physical',
          status: 'completed',
          createdAt: '2021-10-28T14:20:27+00:00',
        },
      ],
    };
    const ctActions = createCtActions(mockedShipmentResponse, mockedCtObject);
    ctActions.forEach(action => {
      expect(action).toMatchSnapshot();
    });
  });
});

describe('updateShipment', () => {
  const mockLogError = jest.fn();
  beforeEach(() => {
    Logger.error = mockLogError;
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should prepare params, call mollie, handle response and return actions', async () => {
    const mockedShipmentParams = {
      shipmentId: 'shp_dhq5l7',
      updateParams: {
        orderId: 'ord_3uwvfd',
        tracking: {
          carrier: 'PostNL',
          code: '3SKABA000000000',
          url: 'http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C',
        },
      },
    };
    const mockedCtObject = {
      custom: {
        fields: {
          updateShipmentRequest: '{"shipmentId":"shp_dhq5l7","tracking":{"carrier":"PostNL","code":"3SKABA000000000","url":"http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C"}}',
        },
      },
    };
    const mockedShipmentResponse: any = {
      resource: 'shipment',
      id: 'shp_dhq5l7',
      orderId: 'ord_3uwvfd',
      createdAt: '2021-10-28T14:30:25+00:00',
      tracking: {
        carrier: 'PostNL',
        code: '3SKABA000000000',
        url: 'http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C',
      },
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.juguf1',
          orderId: 'ord_3uwvfd',
          name: 'orange',
          sku: null,
          type: 'physical',
          status: 'completed',
          createdAt: '2021-10-28T14:20:27+00:00',
        },
      ],
    };
    const mockedCtActions: Action[] = [];
    const getShipmentParams = jest.fn().mockResolvedValueOnce(mockedShipmentParams);
    const createCtActions = jest.fn().mockReturnValueOnce(mockedCtActions);
    const mollieClient = { orders_shipments: { update: jest.fn().mockResolvedValueOnce(mockedShipmentResponse) } } as any;

    const updateShipmentRes = await updateShipment(mockedCtObject, mollieClient, getShipmentParams, createCtActions);
    expect(getShipmentParams).toBeCalledWith(mockedCtObject);
    expect(mollieClient.orders_shipments.update).toHaveBeenCalledWith(mockedShipmentParams.shipmentId, mockedShipmentParams.updateParams);
    expect(createCtActions).toBeCalledWith(mockedShipmentResponse, mockedCtObject);
    expect(updateShipmentRes.status).toBe(200);
  });
  it('Should return commercetools formated error if one of the functions fails', async () => {
    const mockedError = { status: 400, title: 'Could not make parameters needed to update Mollie shipment.', field: 'updateShipmentRequest' };
    const getShipmentParams = jest.fn().mockRejectedValueOnce(mockedError);
    const createCtActions = jest.fn();
    const mollieClient = { orders_shipments: { update: jest.fn() } } as any;

    const updateShipmentRes = await updateShipment({}, mollieClient, getShipmentParams, createCtActions);
    expect(updateShipmentRes.status).toBe(400);
    expect(updateShipmentRes.errors).toHaveLength(1);
  });
});
