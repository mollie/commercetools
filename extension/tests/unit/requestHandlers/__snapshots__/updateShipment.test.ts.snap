// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCtActions Should create correct ct actions from request and mollies payment 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "updateShipment",
    "createdAt": "2021-10-08T12:12:02.625Z",
    "request": "{\\"shipmentId\\":\\"shp_dhq5l7\\",\\"tracking\\":{\\"carrier\\":\\"PostNL\\",\\"code\\":\\"3SKABA000000000\\",\\"url\\":\\"http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C\\"}}",
    "response": "{\\"resource\\":\\"shipment\\",\\"id\\":\\"shp_dhq5l7\\",\\"orderId\\":\\"ord_3uwvfd\\",\\"createdAt\\":\\"2021-10-28T14:30:25+00:00\\",\\"tracking\\":{\\"carrier\\":\\"PostNL\\",\\"code\\":\\"3SKABA000000000\\",\\"url\\":\\"http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C\\"},\\"lines\\":[{\\"resource\\":\\"orderline\\",\\"id\\":\\"odl_1.juguf1\\",\\"orderId\\":\\"ord_3uwvfd\\",\\"name\\":\\"orange\\",\\"sku\\":null,\\"type\\":\\"physical\\",\\"status\\":\\"completed\\",\\"createdAt\\":\\"2021-10-28T14:20:27+00:00\\"}]}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment 2`] = `
Object {
  "action": "setCustomField",
  "name": "updateShipmentResponse",
  "value": "{\\"resource\\":\\"shipment\\",\\"id\\":\\"shp_dhq5l7\\",\\"orderId\\":\\"ord_3uwvfd\\",\\"createdAt\\":\\"2021-10-28T14:30:25+00:00\\",\\"tracking\\":{\\"carrier\\":\\"PostNL\\",\\"code\\":\\"3SKABA000000000\\",\\"url\\":\\"http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1015CW&D=NL&T=C\\"},\\"lines\\":[{\\"resource\\":\\"orderline\\",\\"id\\":\\"odl_1.juguf1\\",\\"orderId\\":\\"ord_3uwvfd\\",\\"name\\":\\"orange\\",\\"sku\\":null,\\"type\\":\\"physical\\",\\"status\\":\\"completed\\",\\"createdAt\\":\\"2021-10-28T14:20:27+00:00\\"}]}",
}
`;
